packages/expo-router/src/fork/useLinking.ts

```diff
@@ -1,3 +1,405 @@
+    queue = queue.then(cb);
+  };
+  return callback;
+};
+
+let linkingHandlers: Symbol[] = [];
+
+type Options = LinkingOptions<ParamListBase> & {
+  independent?: boolean;
+};
+
+export default function useLinking(
+  ref: React.RefObject<NavigationContainerRef<ParamListBase>>,
+  {
+    independent,
+    enabled = true,
+    config,
+    getStateFromPath = getStateFromPathDefault,
+    getPathFromState = getPathFromStateDefault,
+    getActionFromState = getActionFromStateDefault,
+  }: Options
+) {
+  React.useEffect(() => {
+    if (process.env.NODE_ENV === 'production') {
+      return undefined;
+    }
+
+    if (independent) {
+      return undefined;
+    }
+
+    if (enabled !== false && linkingHandlers.length) {
+      console.error(
+        [
+          'Looks like you have configured linking in multiple places. This is likely an error since deep links should only be handled in one place to avoid conflicts. Make sure that:',
+          "- You don't have multiple NavigationContainers in the app each with 'linking' enabled",
+          '- Only a single instance of the root component is rendered',
+        ]
+          .join('\n')
+          .trim()
+      );
+    }
+
+    const handler = Symbol();
+
+    if (enabled !== false) {
+      linkingHandlers.push(handler);
+    }
+
+    return () => {
+      const index = linkingHandlers.indexOf(handler);
+
+      if (index > -1) {
+        linkingHandlers.splice(index, 1);
+      }
+    };
+  }, [enabled, independent]);
+
+  const [history] = React.useState(createMemoryHistory);
+
+  // We store these options in ref to avoid re-creating getInitialState and re-subscribing listeners
+  // This lets user avoid wrapping the items in `React.useCallback` or `React.useMemo`
+  // Not re-creating `getInitialState` is important coz it makes it easier for the user to use in an effect
+  const enabledRef = React.useRef(enabled);
+  const configRef = React.useRef(config);
+  const getStateFromPathRef = React.useRef(getStateFromPath);
+  const getPathFromStateRef = React.useRef(getPathFromState);
+  const getActionFromStateRef = React.useRef(getActionFromState);
+
+  React.useEffect(() => {
+    enabledRef.current = enabled;
+    configRef.current = config;
+    getStateFromPathRef.current = getStateFromPath;
+    getPathFromStateRef.current = getPathFromState;
+    getActionFromStateRef.current = getActionFromState;
+  });
+
+  /* Start of fork. Source: https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/useLinking.tsx#L142 */
+  // ServerContext is used inside ServerContainer to set the location during SSR: https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/ServerContainer.tsx#L50-L54
+  // Expo Router uses the `initialLocation` prop to set the initial location during SSR:
+  const location = React.useContext(ServerLocationContext);
+  const server = { location };
```

Comment:
@EvanBacon I'm not sure this behaviour is correct.

`ServerContext` is only used in React Navigation inside the `ServerContainer` to set the initial URL https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/ServerContainer.tsx#L50-L54

But Expo Router never sets the location during `renderToStaticContent` 
https://github.com/expo/expo/blob/01a47a80611582494484ee6954d6f5ce6a860131/packages/expo-router/src/static/renderStaticContent.tsx#L108

If I were to replicate this current behaviour exactly, I should just remove `ServerContext`. Instead I've replaced it with a new context (the Provider sets the location in ExpoRoot)